### PR TITLE
Gemma-4 E2B-it on JAX path (PLE + KV-share + double-wide MLP)

### PR DESCRIPTION
**Pull request by QiliangCui2023**
## Summary

Adds JAX-native support for `Gemma4ForConditionalGeneration` text-only inference (E2B-it tested; the E-family architecture extends to E4B with the same surface).

Architectural extensions on top of the standard Gemma3-style stack:

- **Per-Layer Embeddings (PLE)** — per-layer scaled embeddings projected and combined with text embeds via the vllm-pytorch reference.
- **KV-share** — the last `num_kv_shared_layers` layers reuse the K/V cache slot of the last preceding same-type layer (sliding vs full). Wired via `kv_sharing_target_layer_name` on the Attention module + the kv_cache_manager's `shared_kv_cache_layers` redirect.
- **Double-wide MLP** — KV-shared layers get `2 * intermediate_size` when `text_config.use_double_wide_mlp` is set.
- **Weight loader** — PLE weights (`embed_tokens_per_layer`, `per_layer_model_projection`, `per_layer_projection_norm`); skip `audio_tower` / activation-calibration scalars; full Q/K/V load for shared layers (K/V is computed but the kernel skips the cache write).

### `tpu_inference/models/common/kv_share.py` (new)

`compute_kv_share_map(text_config)` — generic helper that derives `{shared_idx: source_idx}` from HF `text_config.num_kv_shared_layers` + `layer_types`. The implementation operates only on HF text-config attributes (no model-implementation dependencies), so `runner/kv_cache_manager.py` can call it without pulling in any actual model code. Reused inside `gemma4.py` itself for the per-layer `is_kv_shared_layer` decision. Any future JAX model whose HF config uses the same convention picks it up automatically.

## Test plan

- `tests/models/jax/test_gemma4_ple.py` — PLE structural + numerical tests.
- `tests/models/jax/test_gemma4_kv_share.py` — `compute_kv_share_map` edge cases + KV-share layer structure.
- `tests/models/jax/test_gemma4.py` — extended for E2B.
- `.buildkite/models/google_gemma-4-E2B-it.yml` — nightly serving check on v6e (TP=1, chat-templated offline_inference + chat-templated gsm8k + text-only `vllm bench serve`).

Manual coherence verification (offline_inference on v6e single chip, chat-templated):

| Prompt | Generated |
|---|---|
| Hello, my name is | "Hello! It's nice to meet you. What is your name? 😊" |
| The capital of France is | "**Paris**." |
| The largest planet in our solar system is | "**Jupiter**." |
| Who painted the ceiling of the Sistine Chapel? | "**Michelangelo**." |

## Performance & accuracy

The yml ships with **default RPA v3** as the production gate (fully deterministic gsm8k across runs). Batched RPA is opt-in — add `USE_BATCHED_RPA_KERNEL: "1"` to the step env block. Both kernels were verified equivalent across dev-pipeline runs.

**Builds:** [BK #459](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/459) · [#460](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/460) · [#461](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/461) · [#462](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/462) · [#463](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/463)

### gsm8k (chat-templated, 1319 prompts, in-process `lm_eval --model vllm`)

| filter | default RPA | batched RPA |
|---|---|---|
| `exact_match,strict-match` | **0.5375** | **0.5284** |
| `exact_match,flexible-extract` | **0.6846** | **0.6816** |

Default RPA is fully deterministic across runs. Batched RPA jitters by ~1 pp due to FP-tie flips in the alternate attention path (the "drafted by" vs "written by" Declaration-of-Independence case observed in [BK #458](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/458) manual comparison).

### Throughput (`vllm bench serve`, dataset=random, 1024-in / 128-out, 128 prompts, v6e TP=1, `max_num_seqs=64`, `gpu_memory_utilization=0.94`)

| metric | default RPA | batched RPA |
|---|---|---|
| Request throughput (req/s) | **20.84** ([#461](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/461): 20.37, [#462](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/462): 21.31) | **17.51** ([#462](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/462)) |
| Output throughput (tok/s) | **2667** ([#461](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/461): 2607, [#462](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/462): 2727) | **2241** ([#462](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/462)) |
| Mean TTFT (ms) | 2208 | 2466 |
| Mean TPOT (ms) | 19.97 | 24.24 |

Batched is ~15–18 % slower on this workload on v6e (consistent with the 35-prompt smoke timings in [BK #458](https://buildkite.com/tpu-commons/tpu-inference-dev/builds/458)).

## Depends on

- **#2601** — RPA v3 kernel `update_kv_cache=False` support. The KV-shared layers need the kernel to skip cache writes via this flag; without it, shared layers overwrite the source layer's cache slot.
- **#2602** — `kv_cache_manager` None-handling for `text_config` attributes. E2B's HF config has `num_global_key_value_heads=None`, which crashes spec generation without that fix.
- **#2632** (merged as `19b2af49`) — restores #2601 (which was reverted via #2628) and extends `update_kv_cache=False` to the experimental batched RPA kernel.

## What's currently text-only

This PR does not bring up Gemma-4 vision or audio. The weight loader explicitly skips `audio_tower` / `embed_audio`; the vision tower is loaded but not wired through `Gemma4ForConditionalGeneration.compute_hidden_states` for image inputs. Multimodal bringup is a separate follow-up.

### CODE REVIEW

File: tests/kernels/ragged_paged_attention_kernel_v3_kv_share_test.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
+# Copyright 2026 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (QiliangCui2023) [inline tests/kernels/ragged_paged_attention_kernel_v3_kv_share_test.py]: instead of writing a new file for testing, merge the test to existing RPA v3 kernel test file.

File: .buildkite/pipeline_dev.yml [hunk 2]
```diff
@@ -40,6 +40,31 @@ steps:
   # -----------------------------------------------------------------
   # Build Step (required by all demo steps below)
   # -----------------------------------------------------------------
+  - label: ":docker: Build and Push Base Image (tpu6e)"
```
Comment (QiliangCui2023) [inline .buildkite/pipeline_dev.yml]: this file is used for quick running test. this will be removed and won't be merged.

File: tpu_inference/runner/kv_cache_manager.py [hunk 3]
```diff
@@ -451,7 +451,21 @@ def get_kv_cache_spec(self):
             base_num_kv_heads = model_config.get_total_num_kv_heads()
             base_head_size = model_config.get_head_size()
 
+            # KV-share (JAX path): for the JAX-path (no
+            # vllm-side Attention modules to scan), populate the shared
+            # mapping directly from text_config via the shared helper.
+            from tpu_inference.models.jax.gemma4 import gemma4_kv_share_map
```
Comment (lk-chen) [inline tpu_inference/runner/kv_cache_manager.py]: let's try to avoid model-specific logic in kv_cache_manager

File: tpu_inference/models/jax/gemma4.py [hunk 4]
```diff
@@ -58,9 +101,16 @@ def __init__(self,
...
+                 prefix: str = "",
+                 intermediate_size_override: Optional[int] = None):
         hidden_size = config.hidden_size
-        intermediate_size = config.intermediate_size
+        # Double-wide MLP support (vllm reference at
+        # vllm/model_executor/models/gemma4.py:599-610). When set,
+        # intermediate_size_override replaces config.intermediate_size for
+        # this layer — used by KV-shared layers when use_double_wide_mlp=True.
+        intermediate_size = (intermediate_size_override
+                             if intermediate_size_override is not None else
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py]: there's only 1 caller, and this _override is never None

**Review by lk-chen — COMMENTED** [review]
still in the middle of review, but I think at least we should separate
* kernel change
* kv_cache_manager change

Into dedicated PRs

**Comment by QiliangCui** [comment]
@lk-chen thanks for the review. I've split the PR per your request:

- **#2601** — kernel `update_kv_cache=False` support (kernel.py + interface plumbing + folded kernel tests).
- **#2602** — `kv_cache_manager` None-handling for `text_config` attributes.
- **This PR (#2572)** — trimmed to the Gemma-4 model only.

For your other two inline comments:

- `kv_cache_manager.py:457` (avoid model-specific logic) — moved the helper out to `tpu_inference/models/common/kv_share.py::compute_kv_share_map`. The runner no longer imports from any specific model file; it derives the mapping generically from HF `text_config` attributes. Any future JAX model that uses the same `num_kv_shared_layers` + `layer_types` convention picks it up automatically.
- `gemma4.py:112` (`intermediate_size_override` never None) — dropped the Optional default and the conditional fallback. The single caller now passes `intermediate_size` directly.

This PR's BK step will go red until #2601 and #2602 merge. Functional verification of the combined diff (correct E2B chat-templated output across 35 prompts on v6e) was done end-to-end before the split.

File: tpu_inference/models/jax/gemma4.py [hunk 5]
```diff
@@ -695,6 +781,14 @@ def __init__(self,
         # Gemma 4: Embeddings are scaled by sqrt(hidden_size)
         self.embedding_scale = hidden_size**0.5
 
+        # PLE (Per-Layer Embedding) — model-level modules.
+        # Active when hidden_size_per_layer_input > 0 (E2B/E4B).
+        self.hidden_size_per_layer_input = getattr(
+            text_config, "hidden_size_per_layer_input", 0) or 0
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py]: is this "or 0" necessary?

File: tpu_inference/models/jax/gemma4.py [hunk 6]
```diff
@@ -521,6 +540,11 @@ def __init__(self,
 
         self.is_sliding = self.layer_type == "sliding_attention"
 
+        # PLE (Per-Layer Embedding) — per-layer modules in the decoder.
+        # Active when hidden_size_per_layer_input > 0 (E2B/E4B: 256; 26B/31B: 0).
+        self.hidden_size_per_layer_input = getattr(
+            text_config, "hidden_size_per_layer_input", 0) or 0
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py]: ditto, is "or 0" necessary?

File: tpu_inference/models/jax/gemma4.py [hunk 7]
```diff
@@ -365,8 +354,8 @@ def __init__(self,
             bias_shape=(self.num_kv_heads,
                         self.head_dim) if config.attention_bias else None,
             param_dtype=dtype,
-            kernel_init=nnx.with_partitioning(init_fn, _kv_kernel_spec),
-            bias_init=nnx.with_partitioning(init_fn, _kv_bias_spec)
+            kernel_init=nnx.with_partitioning(init_fn, (None, None, "model")),
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py]: this must be unintended revert

File: tpu_inference/models/jax/gemma4.py [hunk 8]
```diff
@@ -709,6 +803,59 @@ def __init__(self,
...
+            # at use site).
+            self.embed_tokens_per_layer = JaxEmbed(
+                num_embeddings=self.vocab_size_per_layer_input,
+                features=L * P,
+                param_dtype=dtype,
+                embedding_init=nnx.with_partitioning(init_fn, (None, None)),
+                rngs=rng,
+                quant_config=vllm_config.quant_config,
+                prefix=prefix + ".embed_tokens_per_layer",
+            )
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py:840]: You can set "weight_loader" here, like
```
set_attr(self.embed_tokens_per_layer.weight, "weight_loader", ...)
```

File: tpu_inference/models/jax/utils/weight_utils.py [hunk 9]
```diff
@@ -909,6 +909,10 @@ def __init__(self,
                     permute_dims = (1, 2, 0)
                 elif "embed_tokens.weight" in name:
                     permute_dims = (0, 1)
+                elif "embed_tokens_per_layer.weight" in name:
```
Comment (lk-chen) [inline tpu_inference/models/jax/utils/weight_utils.py]: plz check the comments at line 883

see https://github.com/vllm-project/tpu-inference/pull/2572/changes#r3244239291

File: tpu_inference/runner/kv_cache_manager.py [hunk 10]
```diff
@@ -451,7 +451,25 @@ def get_kv_cache_spec(self):
             base_num_kv_heads = model_config.get_total_num_kv_heads()
             base_head_size = model_config.get_head_size()
 
+            # KV-share (JAX path): vllm-side models declare KV-share via
+            # per-Attention `kv_sharing_target_layer_name`; the JAX-side
+            # models have no equivalent runtime declaration. Derive the
+            # same mapping from the HF text_config attributes. Returns {}
+            # for models that don't use KV-share, so this is a no-op for
+            # the common case.
+            from tpu_inference.models.common.kv_share import \
```
Comment (lk-chen) [inline tpu_inference/runner/kv_cache_manager.py]: may want to move import to top of file

File: tests/models/jax/test_gemma4.py [hunk 11]
```diff
@@ -32,10 +32,18 @@
...
-    @pytest.mark.parametrize("model_name", [
-        "google/gemma-4-31B-it",
-        "google/gemma-4-26B-A4B-it",
-    ])
+    @pytest.mark.parametrize(
+        "model_name",
+        [
+            "google/gemma-4-31B-it",
+            "google/gemma-4-26B-A4B-it",
+            # NOTE: E2B intentionally NOT here. The test truncates to 4 layers
```
Comment (lk-chen) [inline tests/models/jax/test_gemma4.py]: you can 
1. have another test case with world_size = 1, or
2. skip test if world_size = 1 and model is E2B

**Comment by QiliangCui2023** [comment]
## Parking this PR

Marking as draft until the v3 RPA kernel `update_kv_cache` feature re-lands.

**Reason:** #2601 (kernel feature this PR depends on) was reverted via #2628 due to a performance regression on the common non-KV-shared path. Without the kernel flag, this PR's KV-shared layers would silently corrupt the cache — so it can't merge as-is.

**Status:**
- No changes needed to this PR's diff. The `attention(..., update_kv_cache=...)` call site is correct; it just requires the kernel-side support that's currently absent on main.
- CI will be red on this branch until the kernel feature re-lands. Not actively maintaining it.

**Next step:** once a fixed version of #2601 merges to main, I'll rebase and flip this PR back to ready-for-review.

File: tpu_inference/models/jax/gemma4.py [hunk 12]
```diff
@@ -324,7 +331,7 @@ def __init__(self,
         # XLA inserting all-to-all reshuffles every layer. When num_kv_heads <
         # tp_size (e.g. global layers with k_eq_v + num_global_key_value_heads
         # = 4 at TP=8), fall back to sharding the head_dim axis; the kernel
-        # replicates kv-heads internally for that case.
+        # replicates kv-heads internally for that case. (Restored from #2585.)
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py]: unintended comment?

**Review by lk-chen — APPROVED** [review]
LGTM with nit
